cirq-core/cirq/devices/grid_device_metadata.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Metadata subtype for 2D Homogenous devices."""
+
+from typing import (
+    TYPE_CHECKING,
+    Optional,
+    FrozenSet,
+    Iterable,
+    Tuple,
+    Dict,
+)
+
+import networkx as nx
+from cirq import value
+from cirq.devices import device
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class GridDeviceMetadata(device.DeviceMetadata):
+    """Hardware metadata for homogenous 2d symmetric grid devices."""
+
+    def __init__(
+        self,
+        qubit_pairs: Iterable[Tuple['cirq.Qid', 'cirq.Qid']],
```

Comment:
Note that there could be isolated qubits that are not coupled to any other qubit. Imagine a real device where we have broken couplers so can't perform 2-qubit gates with a qubit, but still need to bias it to operate the other qubits on the device. This is definitely an edge case, but something we might want to support.